cluster-autoscaler: Re: AWS Autoscaler autodiscover ASG names and sizes

mumoshu opened the pull request:
A redo of https://github.com/kubernetes/contrib/pull/2549

---

This is an alternative implementation of https://github.com/kubernetes/contrib/pull/1982

Notable differences from the original PR are:

* A new flag named `--node-group-auto-discovery` is introduced for opting in to enable the auto-discovery feature.
  * For example, specifying `--cloud-provider aws --node-group-auto-discovery asg:tag=k8s.io/cluster-autoscaler/enabled` instructs CA to auto-discover ASGs tagged with `k8s.io/cluster-autoscaler/enabled` to be used as target node groups
* The new code path introduced by this PR is executed only when `node-group-auto-discovery` is specified. There is relatively less chance to break existing features by introducing this change

Resolves https://github.com/kubernetes/contrib/issues/1956

---

Other notes:

* ~~I thought it might be a good idea to implement this feature on top of the dynamic reconfiguration #2181 initially~~
  * ~~However it turned out it is easier and less verbose to just implement an alternative aws cloud provider like what I've done for this PR~~
* We rely mainly on the `DescribeTags` API rather than `DescribeAutoScalingGroups` so that AWS can filter out unnecessary ASGs which doesn't belong to the k8s cluster, for us.
  * If we relied on `DescribeAutoScalingGroups` here, as it doesn't support `Filter`ing, we'd need to iterate over ALL the ASGs available in an AWS account, which isn't desirable due to unnecessary excessive API calls and network usages

TODO/Possible future improvements before recommending this to everyone:

* Cache the result of an auto-discovery for a configurable period, so that we won't invoke DescribeTags and DescribeAutoScalingGroup APIs too many times

mumoshu commented:
@mwielgus Thanks for maintaining CA 👍
Any chance you could review this?

mumoshu commented:
Hi @MaciekPytel, thanks for maintaining CA!
Would you mind reviewing this?

I'd really like to contribute to CA, especially for the AWS part of it, as a member of SIG-AWS.
However, all I can do after I've submitted a PR is just asking you maintainers for a review!
Sorry for bothering you but I'd greatly appreciate it if you could review this :bow:

mwielgus commented:
@mumoshu - How did you test that this PR works? 

Please update the documentation as well.

mwielgus commented:
cc: @andrewsykim

MaciekPytel commented on cluster-autoscaler/core/autoscaler.go:
```diff
@@ -50,5 +50,8 @@ func NewAutoscaler(opts AutoscalerOptions, predicateChecker *simulator.Predicate
 		configFetcher := dynamic.NewConfigFetcher(opts.ConfigFetcherOptions, kubeClient, kubeEventRecorder)
 		return NewDynamicAutoscaler(autoscalerBuilder, configFetcher)
 	}
+	if opts.NodeGroupAutoDiscovery != "" {
```
If someone passes --configmap and --node-group-auto-discovery the latter will be silently ignored. It would be nice to at least log a warning (and maybe mention it in docs).

mumoshu replied:
Thanks! Added a warning.

mumoshu commented:
@mwielgus First of all - sorry for breaking things in my last PR :bow:

I've updated cloudprovider/aws/README.md to include the new setup.

My testing went like below:
* Created a k8s cluster with kube-aws, added permissions on `DescribeTags` to IAM roles assigned to all the k8s nodes, build my own CA image `mumoshu/cluster-autoscaler:dev3` by running `REGISTRY=mumoshu TAG=dev3 make release` and then deployed it to the cluster
* Added a bunch of `gcr.io/google_containers/hpa-example` pods to make some pending ones.
  * As ASGs had been not tagged yet, I verified the new CA doesn't find any ASG.
* Tagged ASGs with the key specified via the new flag `--node-group-auto-discovery`, in my case `k8s.io/cluster-autoscaler/enabled`
  * As ASGs had been tagged, I verified the new CA does find ASGs and scales them out by increasing descired cap on every ASG

mumoshu commented:
This is the manifest I've used for testing.
Notice I've added `--v=6` so that CA logs which ASGs are auto-discovered.

```yaml
---
apiVersion: extensions/v1beta1
kind: Deployment
metadata:
  name: cluster-autoscaler
  namespace: kube-system
  labels:
    app: cluster-autoscaler
spec:
  replicas: 1
  selector:
    matchLabels:
      app: cluster-autoscaler
  template:
    metadata:
      labels:
        app: cluster-autoscaler
    spec:
      containers:
        - image: mumoshu/cluster-autoscaler:dev3
          name: cluster-autoscaler
          resources:
            limits:
              cpu: 100m
              memory: 300Mi
            requests:
              cpu: 100m
              memory: 300Mi
          command:
            - ./cluster-autoscaler
            - --v=6
            - --stderrthreshold=info
            - --cloud-provider=aws
            - --skip-nodes-with-local-storage=false
            - --expander=least-waste
            - --node-group-auto-discovery=asg:tag=k8s.io/cluster-autoscaler/enabled
          env:
            - name: AWS_REGION
              value: ap-northeast-1
          volumeMounts:
            - name: ssl-certs
              mountPath: /etc/ssl/certs/ca-certificates.crt
              readOnly: true
          imagePullPolicy: "Always"
      volumes:
        - name: ssl-certs
          hostPath:
            path: "/etc/ssl/certs/ca-certificates.crt"
```

andrewsykim commented on cluster-autoscaler/cloudprovider/aws/README.md:
```diff
@@ -131,6 +155,64 @@ spec:
 
+### Auto-Discovery Setup
+
+As of today, docker images including the support for `--node-group-auto-discovery` is not yet published to official repository.
```
As of version "X.X.X"

mumoshu replied:
Sounds good. Thanks for the comment 👍

mumoshu replied:
Addressed in the latest commits

andrewsykim commented on cluster-autoscaler/cloudprovider/aws/README.md:
```diff
@@ -131,6 +155,64 @@ spec:
+As of today, docker images including the support for `--node-group-auto-discovery` is not yet published to official repository.
+Please checkout the latest source of this project locally and run `REGISTRY=<your docker repo> make release` to build and push an image yourself.
+Then, a manifest like below would run a cluster-autoscaler which auto-discovers ASGs tagged with `k8s.io/cluster-autoscaler/enabled` to be node groups.
+Please notice that there're no `--nodes` flags passed to cluster-autoscaler anymore.
```
`s/there're/there are`

andrewsykim replied:
Is there a reason we're not support both the flag and the manifest for at least 1 major version? We should give a deprecation notice for this right?

mumoshu replied:
Thanks 👍 

> Is there a reason we're not support both the flag and the manifest for at least 1 major version?

I'm afraid if I'm following you correctly but the `--nodes` flag is not removed. It's just that you can specify either `--nodes` or `--node-group-auto-discovery`, but not both.

mumoshu replied:
`there're` is fixed in the latest commits

andrewsykim commented on cluster-autoscaler/cloudprovider/aws/README.md:
```diff
@@ -131,6 +155,64 @@ spec:
+          env:
+            - name: AWS_REGION
+              value: us-east-1
+          volumeMounts:
```
I think we can get rid of all the volume mounts for ssl certs if you add
```
RUN  apk add --no-cache --update ca-certificates
```
In the Dockerfile

mumoshu replied:
@andrewsykim Sounds great 👍 
@mwielgus @MaciekPytel Can we add the `ca-certificates` package into the image in an another PR?

mwielgus replied:
yes

mumoshu replied:
Thanks. Opened #38

mumoshu replied:
@andrewsykim FYI ca-certificates are added via #48

andrewsykim commented on cluster-autoscaler/cloudprovider/aws/aws_cloud_provider.go:
```diff
@@ -21,20 +21,74 @@ import (
 	"regexp"
 	"strings"
 
+	"errors"
```
Group stdlib packages with the other stdlib packages please

mumoshu replied:
Thanks! Addressed in the latest commits

andrewsykim commented on cluster-autoscaler/cloudprovider/aws/aws_cloud_provider.go:
```diff
@@ -21,20 +21,74 @@ import (
+	if discoveryOpts.NodeGroupAutoDiscoverySpec != "" {
+		return buildAutoDiscoveringProvider(awsManager, discoveryOpts.NodeGroupAutoDiscoverySpec)
+	}
+	return nil, errors.New("Failed to build an aws cloud provider: Either node group specs or node group auto discovery spec must be specified")
```
What happens if we specify both?

mumoshu replied:
If we allow that in the current implementation, specifying both would result in `--nodes` to be completely ignored.

andrewsykim replied:
Do we want that? Should we validate and/or raise an error?

mumoshu replied:
Perhaps we'd better raise a validation error rather than silently ignoring such mistakes. Would you agree?

andrewsykim replied:
agreed

mumoshu replied:
Thanks! Addressed in the latest commits.

andrewsykim commented on cluster-autoscaler/cloudprovider/aws/aws_cloud_provider.go:
```diff
@@ -21,20 +21,74 @@ import (
+}
+
+func buildAutoDiscoveringProvider(awsManager *AwsManager, spec string) (*autoDiscoveringProvider, error) {
+	tokens := strings.Split(spec, ":")
```
params would be a better name?

mumoshu replied:
I guess no?
`tokens` in `asg:tag=desiredtagkey` are `asg` and `tag=desiredtagkey`.
`param` is a composition of `tag` and `desiredtagkey`, hence the names.

andrewsykim commented on cluster-autoscaler/cloudprovider/aws/aws_cloud_provider.go:
```diff
@@ -21,20 +21,74 @@ import (
+	paramTokens := strings.Split(param, "=")
+	parameterKey := paramTokens[0]
+	if parameterKey != "tag" {
+		return nil, fmt.Errorf("Unsupported parameter key specified for discoverer \"%s\": %s", discoverer, parameterKey)
```
Do we really need the quotes and the `\`?

mumoshu replied:
Not sure myself. I just intended to make it clear that e.g. "asg" is the name of a discoverer.

In other words, `... for (the) discoverer "asg"` seemed clearer for me than `... for (the) discoverer asg`. The latter sounds like a name of asg rather than a discoverer for me.

andrewsykim replied:
But its actually ```... for (the) discoverer: asg```

mumoshu replied:
Thanks for your comment 👍 

I believe it actually look like `Unsupported parameter key specified for discoverer "asg": foobar` when e.g. you've specified `asg:foobar=baz` because the only supported parameter key for the "asg" discoverer is "tag" for now.

Anyway, how about changing this to `Unsupported parameter key "foobar" specified for discoverer "asg". The only supported key is "tag"` for the case so that it would sound more natural IMHO?

BTW I'd welcome nits because my own comments occasionally tend to be nits 😃

mumoshu replied:
Tweaked the message a bit in the latest commit.
Could you tell me WDYT?

andrewsykim commented on cluster-autoscaler/cloudprovider/aws/aws_cloud_provider.go:
```diff
@@ -21,20 +21,74 @@ import (
+	if tag == "" {
+		return nil, errors.New("Invalid ASG tag for auto discovery specified: ASG tag must not be empty")
+	}
+	underlying := &awsCloudProvider{
```
What's the reasoning for calling this `underlying`?

mumoshu replied:
Because it was wrapped inside an autoDiscoveringProvider.
Btw, revisiting the part of code revealed that autoDiscoveringProvider is completely unnecessary!
I've rewritten code around this accordingly.

andrewsykim commented on cluster-autoscaler/cloudprovider/aws/aws_manager.go:
```diff
@@ -24,6 +24,7 @@ import (
 
 	"gopkg.in/gcfg.v1"
 
+	"errors"
```
group with other stdlib packages please

mumoshu replied:
Thanks! Addressed in the latest commits

andrewsykim commented on cluster-autoscaler/cloudprovider/aws/aws_manager.go:
```diff
@@ -35,6 +36,7 @@ import (
 const (
 	operationWaitTimeout  = 5 * time.Second
 	operationPollInterval = 100 * time.Millisecond
+	maxRecords            = 100
```
Should be a bit more specific for what sort of "record" we're limiting here.

mumoshu replied:
Thanks! Addressed in the latest commits

andrewsykim commented on cluster-autoscaler/core/polling_autoscaler_test.go:
```diff
@@ -0,0 +1,43 @@
+
+import (
+	"github.com/stretchr/testify/mock"
+	"testing"
```
stdlib first

mumoshu replied:
Addressed in the latest commits

andrewsykim commented on cluster-autoscaler/main.go:
```diff
@@ -66,6 +66,7 @@ var (
 	cloudConfig             = flag.String("cloud-config", "", "The path to the cloud provider configuration file.  Empty string for no configuration file.")
 	configMapName           = flag.String("configmap", "", "The name of the ConfigMap containing settings used for dynamic reconfiguration. Empty string for no ConfigMap.")
 	namespace               = flag.String("namespace", "kube-system", "Namespace in which cluster-autoscaler run. If a --configmap flag is also provided, ensure that the configmap exists in this namespace before CA runs.")
+	nodeGroupAutoDiscovery  = flag.String("node-group-auto-discovery", "", "One or more definition(s) of node group auto-discovery. A definition is expressed `<name of discoverer per cloud provider>:[<key>[=<value>]]`. For example, specifying `--cloud-provider aws` and `--node-group-auto-discovery asg:tag=cluster-autoscaler/auto-discovery/enabled` resuls in ASGs tagged with `cluster-autoscaler/auto- ...
```
Do we need to document the valid `name of discoverer per cloud provider` and `key`s?

mumoshu replied:
I think so. Where would you like to document it?

andrewsykim replied:
Maybe as part of the flag description?

mumoshu replied:
Thanks! Agreed & Addressed in the latest commits.

mumoshu commented:
Thanks everyone - I've addressed all the review comments.

andrewsykim commented on cluster-autoscaler/cloudprovider/aws/aws_cloud_provider.go:
```diff
@@ -26,15 +27,64 @@ import (
-	aws := &AwsCloudProvider{
+func BuildAwsCloudProvider(awsManager *AwsManager, discoveryOpts cloudprovider.NodeGroupDiscoveryOptions) (cloudprovider.CloudProvider, error) {
+	staticDiscoverySpecified := len(discoveryOpts.NodeGroupSpecs) > 0
+	autoDisvoerySpecified := discoveryOpts.NodeGroupAutoDiscoverySpec != ""
```
`s/autoDisvoerySpecified/autoDiscoverySpecified`

mumoshu replied:
@andrewsykim Good catch 👍 Fixed in the latest commit

andrewsykim commented on cluster-autoscaler/cloudprovider/aws/aws_cloud_provider.go:
```diff
@@ -26,15 +27,64 @@ import (
+	staticDiscoverySpecified := len(discoveryOpts.NodeGroupSpecs) > 0
+	autoDiscoverySpecified := discoveryOpts.NodeGroupAutoDiscoverySpec != ""
+	if staticDiscoverySpecified && autoDiscoverySpecified {
+		return nil, errors.New("Failed to build an aws cloud provider: Either node group specs or node group auto discovery spec can be specified but not both")
```
Can we include the flag values in this error message (`--node-group-auto-discovery` or `--nodes`)

mumoshu replied:
Addressed in the latest commit. Thanks!

andrewsykim commented on cluster-autoscaler/cloudprovider/aws/aws_cloud_provider.go:
```diff
@@ -26,15 +27,64 @@ import (
+func buildAutoDiscoveringProvider(awsManager *AwsManager, spec string) (*awsCloudProvider, error) {
+	tokens := strings.Split(spec, ":")
+	if len(tokens) != 2 {
+		return nil, fmt.Errorf("Invalid node group auto discovery spec specified: %s", spec)
```
Can we mention the flag name here too `--node-group-auto-discovery` for clarity

mumoshu replied:
Addressed in the latest commit. Thanks!

andrewsykim commented on cluster-autoscaler/core/polling_autoscaler.go:
```diff
@@ -0,0 +1,69 @@
+
+// Poll latest data from cloud provider to recreate this autoscaler
+func (a *PollingAutoscaler) Poll() error {
+	// For safety, any config change should stop and recreate all the stuff running in CA hence recreating all the Autoscaler instance here
```
I might be missing something here, but how exactly are we detecting changes to config? At a glance, it seems like we're always recreating the autoscaler from scratch. Is this what you meant by:
> Cache the result of an auto-discovery for a configurable period, so that we won't invoke DescribeTags and DescribeAutoScalingGroup APIs too many times

mumoshu replied:
@andrewsykim Sorry if it wasn't clear but as the sentence is resided in "Possible future improvements before recommending this to everyone:", it is a TODO, possibly addressed in another PR if anyone considers it to be necessary.

> it seems like we're always recreating the autoscaler from scratch

Yes, that's definitely how it works for now.

Would you want the caching behavior before merging this PR?
If not, I can submit an another PR for it anyway!

mumoshu replied:
My assumption here is that we won't want to introduce the caching behavior in this PR because IMHO it requires adding flags like `--node-group-auto-discovery-cache-period` which may further delay finishing this PR

mwielgus commented:
@andrewsykim Thank you for reviewing tins PR.

@mumoshu Please address the remaining comments, confirm that you ran the test on the final code and we are good to merge this PR.

mwielgus commented:
And please squash the commits to no more than 2-3 commits. 11, especially with ambiguous names like "stdlib first", is too much :).

mumoshu commented:
@mwielgus Thanks!

I've addressed all the comments except https://github.com/kubernetes/autoscaler/pull/11#discussion_r114671590, as I believe it should be addressed in an another PR (if necessary).
I can open a new github issue dedicated to that if you'd like me to do so.

I've also squashed all the commits into one, which is what I meant to do after the review is finished.

The travis build is passing and my cluster with the `mumoshu/dev5` image built from the source from this PR is working without any problem so far with the [above configuration](https://github.com/kubernetes/autoscaler/pull/11#issuecomment-297737376)

mumoshu commented:
@mwielgus I've resolved a conflict with the latest master branch and rebase onto it. I'm assuming this is ready to be merged.

mwielgus commented:
LGTM
